### ISSUE TITLE
ANR on Android 9/10 devices

### ISSUE THREAD

**Issue by mbenzenhoefer** [launchdarkly/android-client-sdk#99]
**Describe the bug**
Google play console shows several ANRs (private package name hidden):
```
Broadcast of Intent { flg=0x14 cmp=my.package.name/com.launchdarkly.android.PollingUpdater (has extras) }

com.launchdarkly.android.PollingUpdater
```

**Expected behavior**
No ANR in console.

**Logs**
```
"Thread-7" tid=25 Waiting
"Thread-7" prio=5 tid=25 Waiting
  | group="main" sCount=1 dsCount=0 flags=1 obj=0x13340ce0 self=0x7977182800
  | sysTid=16374 nice=10 cgrp=bg_cached sched=0/0 handle=0x790be2cd50
  | state=S schedstat=( 601500 348923 2 ) utm=0 stm=0 core=1 HZ=100
  | stack=0x790bd2a000-0x790bd2c000 stackSize=1039KB
  | held mutexes=
  at sun.misc.Unsafe.park (Native method)
- waiting on an unknown object
  at java.util.concurrent.locks.LockSupport.park (LockSupport.java:190)
  at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await (AbstractQueuedSynchronizer.java:2067)
  at java.util.concurrent.LinkedBlockingQueue.take (LinkedBlockingQueue.java:442)
  at java.util.concurrent.ThreadPoolExecutor.getTask (ThreadPoolExecutor.java:1092)
  at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1152)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:641)
  at com.launchdarkly.android.BackgroundThreadExecutor$PriorityThreadFactory$1.run (BackgroundThreadExecutor.java:48)
  at java.lang.Thread.run (Thread.java:919)
```
**SDK version**
2.10.0

**Language version, developer tools**
Android with kotlin source code (v1.3.61), error appears on Google Play Console.

**OS/platform**
Android, only occurs on Android 9/10 devices and (if not just occasionally) only Samsung devices.

Thanks for your help!

**Comment by gwhelanLD** [comment]
Hi @mbenzenhoefer,

Thanks for filing this. We'll attempt to reproduce on our end. If you have any more details about the conditions that cause this issue to occur please let us know. If it'd be easier to avoid censoring application packages/etc, feel free to also reach out to support with additional details.

We'll update here when we know more.

Thanks,
@gwhelanLD

**Comment by mbenzenhoefer** [comment]
@gwhelanLD any updates on this issue? We have this issue more than 70 times in Play Console now.
Mainly Android 9/10 devices, but also 7/8.

**Comment by dsuresh-ap** [comment]
We have this same ANR plus a few others but not sure if they are related to LaunchDarkly.
Not sure if it is related but we have this ANR as well:
```
Broadcast of Intent { act=android.intent.action.LOCALE_CHANGED flg=0x11200010 cmp=<package>/.app.locale.LocaleChangeReceiver }, InvisibleToUser
```

**Comment by abhisheksisodia** [comment]
Seeing same issue in our app

```
Broadcast of Intent { flg=0x14 cmp=com.scotiabank.banking/com.launchdarkly.android.PollingUpdater (has extras) }
com.launchdarkly.android.PollingUpdater
```
Affected almost 300 users in past 60 days

**Comment by volo-droid** [comment]
It looks like the ANR happens when LD's `PollingUpdater` broadcast receiver calls `android.net.ConnectivityManager.getActiveNetworkInfo`:

```
"main" prio=5 tid=1 Native

  #00  pc 000000000008023c  /apex/com.android.runtime/lib64/bionic/libc.so (syscall+28)
  #00  pc 000000000014bf34  /apex/com.android.runtime/lib64/libart.so (art::ConditionVariable::WaitHoldingLocks(art::Thread*)+148)
  #00  pc 0000000000589054  /apex/com.android.runtime/lib64/libart.so (art::GoToRunnable(art::Thread*) (.llvm.6398098689801567624)+440)
  #00  pc 0000000000588e58  /apex/com.android.runtime/lib64/libart.so (art::JniMethodEnd(unsigned int, art::Thread*)+28)
  at android.os.BinderProxy.transactNative (Native method)
  at android.os.BinderProxy.transact (BinderProxy.java:575)
  at android.net.IConnectivityManager$Stub$Proxy.getActiveNetworkInfo (IConnectivityManager.java:2992)
  at android.net.ConnectivityManager.getActiveNetworkInfo (ConnectivityManager.java:1120)
  at com.launchdarkly.android.LDUtil.isInternetConnected (LDUtil.java:25)
  at com.launchdarkly.android.LDUtil.isClientConnected (LDUtil.java:37)
  at com.launchdarkly.android.HttpFeatureFlagFetcher.fetch (HttpFeatureFlagFetcher.java:69)
  at com.launchdarkly.android.DefaultUserManager.updateCurrentUser (DefaultUserManager.java:75)
  at com.launchdarkly.android.ConnectivityManager.triggerPoll (ConnectivityManager.java:401)
  at com.launchdarkly.android.LDClient.triggerPoll (LDClient.java:572)
  at com.launchdarkly.android.LDClient.triggerPollInstances (LDClient.java:681)
  at com.launchdarkly.android.PollingUpdater.onReceive (PollingUpdater.java:25)
```

**Comment by girubhai** [comment]
any updates ?

**Comment by volo-droid** [comment]
@gwhelanLD can you please look on the stacktrace I posted above? 
https://github.com/launchdarkly/android-client-sdk/issues/99#issuecomment-727270589

It looks like the ANR is caused by `android.net.ConnectivityManager.getActiveNetworkInfo` method. The method is [deprecated starting from API 29](https://developer.android.com/reference/android/net/NetworkInfo) and Google suggests to use `ConnectivityManager.NetworkCallback` API.

**Comment by dsuresh-ap** [comment]
Not sure if this is related but we are also seeing ANRs with the following blocked thread:
`Input dispatching timed out`

```
main" prio=5 tid=1 Blocked
  at com.launchdarkly.android.LDClient.onNetworkConnectivityChangeInstances (LDClient.java)
  at com.launchdarkly.android.ConnectivityReceiver.onReceive (ConnectivityReceiver.java:29)
  at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0$LoadedApk$ReceiverDispatcher$Args (LoadedApk.java:1560)
  at android.app.-$$Lambda$LoadedApk$ReceiverDispatcher$Args$_BumDX2UKsnxLVrE6UJsJZkotuA.run (lambda)
  at android.os.Handler.handleCallback (Handler.java:883)
  at android.os.Handler.dispatchMessage (Handler.java:100)
  at android.os.Looper.loop (Looper.java:214)
  at android.app.ActivityThread.main (ActivityThread.java:7710)
  at java.lang.reflect.Method.invoke (Method.java)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:516)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:950)
```

**Comment by erawhctim** [comment]
> It looks like the ANR is caused by android.net.ConnectivityManager.getActiveNetworkInfo method. The method is deprecated starting from API 29 and Google suggests to use ConnectivityManager.NetworkCallback API.

@gwhelanLD & team, were you ever able to reproduce this? Or any update since these last comments? Curious if the `ConnectivityManager` deprecation was indeed the root cause, or if the ANRs reported here were due to some other application-side error (outside of the LD SDK). 

I know this was a while ago, so my apologies for digging this back up. My team is trying to pinpoint the source of some LD-related ANRs that we've seen recently, so any extra context/info would be helpful

**Comment by ssaxena01** [comment]
Broadcast of Intent { flg=0x10 cmp=com.onemedical.android/com.launchdarkly.sdk.android.PollingUpdater }

We've seen 681  ANRs in the past 30 days. This includes Android 9/10/11/12. Any updates on this issue? I'm on LD SDK v 3.0.2.

[gwhelanLD](https://github.com/gwhelanLD)

**Comment by aengelberg** [comment]
Hey everyone, apologies this ticket has sat for so long. To address the ANRs, we've been auditing the codebase for unnecessary locking or I/O blockage that could cause the main UI thread to hang.

Today, we released version 3.1.6 which fixes some excessive locking that we saw in `LDClient`. A bunch of methods, including `triggerPollInstances`, had the `synchronized` keyword, resulting in [lock contention](https://github.com/launchdarkly/android-client-sdk/issues/142) from the main thread, especially when an `LDClient.init()` was in progress at the time these methods were called.

If you are seeing ANRs stem from `PollingUpdater`, please give the latest version a try and let us know if it addresses the problem.

**Comment by louis-launchdarkly** [comment]
As @aengelberg mentioned, several fixes in the 3.1.6 release were related to potential causes of ANRs. We're closing this issue due to a lack of further reports, but please feel free to open a new issue if you are still seeing similar behavior.
